service/stream/exec_data_streamer.go

```diff
@@ -0,0 +1,168 @@
+package stream
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"cloud.google.com/go/storage"
+	"github.com/hashicorp/go-multierror"
+	"github.com/onflow/flow-archive/models/archive"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/engine/execution/computation/computer/uploader"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/grpcutils"
+	"github.com/onflow/flow/protobuf/go/flow/access"
+	execData "github.com/onflow/flow/protobuf/go/flow/executiondata"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type ExecDataStreamer struct {
+	log       zerolog.Logger
+	execApi   execData.ExecutionDataAPIClient
+	accessApi access.AccessAPIClient
+	queue     *archive.SafeDeque // queue of block identifiers for next downloads
+	buffer    *archive.SafeDeque // queue of downloaded execution data records
+	limit     uint               // buffer size limit for downloaded records
+	busy      uint32             // used as a guard to avoid concurrent polling
+	ctx       context.Context
+	chain     flow.ChainID
+}
+
+func NewExecDataStreamer(log zerolog.Logger, accessAddr string, options ...Option) *ExecDataStreamer {
+	cfg := DefaultConfig
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	ctx := context.Background()
+
+	// initialize clients
+	opts := []grpc.DialOption{grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
```

Comment:
This should be configurable. `grpcutils.DefaultMaxMsgSize` most likely won't be enough for large blocks.